src/js/common/validate.js

```diff
@@ -0,0 +1,192 @@
+// extra array element when we run them through `split`.
+gpii.schema.validator.removeEmptySegments = function (el) { return el !== undefined && el !== null && el.length > 0; };
+
+/*
+  Z-Schema represents validation failures using path notation like `#/category/subcategory/field`.  This function
+  converts that notation into a series of path segments that can be passed to `gpii.schema.validator.saveToPath`.
+
+  In addition, this function unifies the two pieces of the z-schema path, `error.params` and `error.path`.  This enables
+  us to represent a hierarchy of validation errors.
+ */
+gpii.schema.validator.extractPathSegments = function (error) {
+    var segments = error.path.split("/").filter(gpii.schema.validator.removeEmptySegments).slice(1);
+    if (error.code === "OBJECT_MISSING_REQUIRED_PROPERTY" && error.params) { segments = segments.concat(error.params); }
+    return segments;
+};
+
+gpii.schema.validator.resolveTargetFromPath = function (target, path) {
+    var resolvedTarget = target;
+    for (var a = 0; a < path.length; a++) {
+        var segment = path[a];
+        if (!resolvedTarget[segment]) {
+            resolvedTarget[segment] = a === path.length - 1 ? [] : {};
+        }
+
+        resolvedTarget = resolvedTarget[segment];
+    }
+    return resolvedTarget;
+};
+
+/*
+  A function to save error messages to the right position within our "sanitized" JSON output (see header for example).
+
+  `path`:
+    An array of path segments pointing to a variable in `errorMap`.  If you are describing validation errors related to
+    the field `category.subcategory.field`, the path would look like `["category", "subcategory", "field"]`.
+
+  `errorString`:
+    An error message to be saved to the right `path` within `errorMap`
+
+  `errorMap`:
+    A map of error messages for the whole document.  This will be modified with new values.
+ */
+gpii.schema.validator.saveToPath = function (path, errorString, errorMap) {
+    var target = errorMap.fieldErrors;
+    // If we have path data, the error is related to a specific field.
+    if (path && path.length > 0) {
+        target = gpii.schema.validator.resolveTargetFromPath(target, path);
+    }
+    // Otherwise, we are working with a document-level error.
+    else {
+        target = errorMap.documentErrors;
+    }
+    target.push(errorString);
+};
+
+fluid.defaults("gpii.schema.validator", {
+    gradeNames: ["fluid.component"],
+    zSchemaOptions: {
+        noExtraKeywords:   true,
+        // This option does not appear to be respected in our version of z-schema, but we set it explicitly to avoid
+        // problems when we upgrade.
+        reportPathAsArray: false
+    },
+    members: {
+        schemaContents: {}
+    },
+    invokers: {
+        validate: {
+            funcName: "gpii.schema.validator.validate",
+            args:     ["{that}", "{arguments}.0", "{arguments}.1"]
+        },
+        getValidator: {
+            funcName: "fluid.fail",
```

Comment:
This can be fluid.notImplemented if you are on current Infusion: https://issues.fluidproject.org/browse/FLUID-5733